File: src/chapters/nonlinear-equations/content/code/false-position.go
package main

import (
	"fmt"
	"math"
)

// Regula falsi (false position) for f on a bracket [a, b].
func falsePosition(f func(float64) float64, a, b, tol float64, maxIter int) float64 {
	fa, fb, c := f(a), f(b), a
	for k := 0; k < maxIter; k++ {
		c = (a*fb - b*fa) / (fb - fa)
		fc := f(c)
		if math.Abs(fc) < tol {
			return c
		}
		if fa*fc < 0 {
			b, fb = c, fc
		} else {
			a, fa = c, fc
		}
	}
	return c
}

func main() {
	fmt.Println(falsePosition(func(x float64) float64 { return x*x - 2 }, 1, 2, 1e-12, 200))
}

File: src/chapters/nonlinear-equations/content/code/false-position.js
// Regula falsi (false position) for f on a bracket [a, b].
function falsePosition(f, a, b, tol = 1e-12, maxIter = 200) {
  let fa = f(a), fb = f(b), c = a;
  for (let k = 0; k < maxIter; k++) {
    c = (a * fb - b * fa) / (fb - fa);
    const fc = f(c);
    if (Math.abs(fc) < tol) return c;
    if (fa * fc < 0) { b = c; fb = fc; } else { a = c; fa = fc; }
  }
  return c;
}
console.log(falsePosition((x) => x * x - 2, 1, 2));
